CMake plumbing to build the loader

illwieckz opened the pull request:
I added some `CMakeLists.tx` code to rebuild `sel_ldr` and `nacl_helper_bootstrap`.

This relies on some unified `DaemonPlatform` framework copied from https://github.com/DaemonEngine/Daemon/pull/1641 on purpose to give this CMake code the same easiness at doing cross-compiled builds.

I need help to complete the `src/trusted/service_runtime/CMakeLists.txt` file.

The `CMakeLists.txt` files are a rewrite of the file `SConstruct` and `*.scons` files with all unit tests deleted. Remaining unported code is commented out with lines starting with `#TODO:`.

Build status:

system|arch|build `sel_ldr`|build `nacl_helper_bootstrap`|run `helloworld` nexe
-|-|-|-|-
windows|amd64|not tested|N/A|not tested
windows|i686|not tested|N/A|not tested
mingw|amd64|✅️|N/A|not tested
mingw|i686|✅️|N/A|not tested
linux|amd64|✅️|✅️|✅️
linux|i686|✅️|✅️|not tested
linux|armhf|✅️|✅️|not tested
linux|armhf 16k|✅️|✅️|not tested
linux|armel|✅️|✅️|not tested
linux|mips|✅️|✅️|not tested
android|armel|✅️|✅️|❌️
macos|amd64|✅️|N/A|not tested

Dynamically linked loader:

system|arch|shared link `sel_ldr`|run `helloworld` nexe
-|-|-|-
windows|amd64|not tested|not tested
windows|i686|not tested|not tested
mingw|amd64|✅️|not tested
mingw|i686|✅️|not tested
linux|amd64|✅️|✅️
linux|i686|✅️|not tested
linux|armhf|✅️|✅️
linux|armhf 16k|✅️|not tested
linux|armel|✅️|not tested
linux|mips|✅️|not tested
android|armel|✅️|not tested
macos|amd64|✅️|not tested

~~Statically linked loader:~~ NOW REMOVED.

system|arch|static link `sel_ldr`|run `helloworld` nexe
-|-|-|-
windows|amd64|not implemented|not implemented
windows|i686|not implemented|not implemented
mingw|amd64|✅️|not tested
mingw|i686|✅️|not tested
linux|amd64|✅️|✅️
linux|i686|✅️|not tested
linux|armhf|✅️|❌️ segfault
linux|armhf 16k|✅️|not tested
linux|armel|✅️|not tested
linux|mips|✅️|not tested
android|armel|✅️|not tested
macos|amd64|✅️|not tested

Things like `linux-mips` were tested only because I ported the SCons code to CMake for completeness and make sure I forgot nothing.

I tested `android-armel` because I remember that in the past @cu-kai tried to get `daemon-tty` running on Android to get a console for his server.

slipher committed:
Allow using system scons instead of chromium one

slipher committed:
Make sel_ldr buildable without Chromium stuff

Only for amd64 host and target. nacl_helper_bootstrap also builds.

- Check in linux_syscall_support.h (like we did for Breakpad)
- Add a hacky symlink so we don't have to include the repository's
  parent directory or name it "native_client"
- Use LLVM from /usr/bin instead of Chromium toolchains
- Remove sysroot flags (depend on host system instead)
- Disable setup of NaCl-target toolchains in scons scripts

slipher committed:
Init from upstream 78751ab2bd2918cdbe4d849412463351088d2e90

illwieckz committed:
Silence Python invalid escape sequence warnings by turning some strings into raw strings

slipher committed:
Try to stop Clang 17+ warn on undefining __STDC_HOSTED__

slipher committed:
Python 3 fixes for some tests

slipher committed:
Avoid depending on another NaCl toolchain for gdb

Disable GDB tests by default and when they are disabled don't try to
look for a 2nd NaCl toolchain just to grab its GDB.

GDB tests on Linux are broken for me anyway in upstream with the
incompatibly old curses lib dependency issue.

slipher committed:
Re-enable NaCl toolchains when requested

With some minor fixes, the build is able to run and produce an irt_core
binary when configured with the Saigo toolchain.
Added documentation how to do this to the README.

Still keeping the 'if nacltools' guards I added so that sel_ldr can be
built without acquiring any NaCl toolchains.

illwieckz committed:
cmake: add the Yokai CMake framework

illwieckz committed:
cmake: add the DaemonPlatform CMake framework

illwieckz committed:
cmake: add MinGW CMake toolchains for cross-compilation

illwieckz committed:
cmake: add MinGW CMake toolchains for cross-compilation

illwieckz committed:
cmake: add CMake scripts to build sel_ldr and nacl_helper_bootstrap

illwieckz committed:
cmake: add CMake scripts to build sel_ldr and nacl_helper_bootstrap

slipher commented:
Is it really so bad to use Scons? Unlike CMake, it can handle multiple toolchains, which makes it well-suited for this repo. Also if we keep the same build system, we can easily compare things between our version and upstream.

illwieckz commented:
> Is it really so bad to use Scons?

Yes. 😅️

> Unlike CMake, it can handle multiple toolchains, which makes it well-suited for this repo.

Uh, totally not. The Scons scripts are not compatible with cross-compiling to begin with.

First purpose of this effort is to be able to use multiple toolchains.

slipher commented:
> The Scons scripts are not compatible with cross-compiling to begin with.

Wrong. With Chromium `native_client` I can do `./scons --mode=nacl,opt-linux platform=x86-32 sel_ldr irt_core_raw` or `./scons --mode=nacl,opt-linux platform=arm sel_ldr irt_core_raw` and everything completes successfully and produces executables of the expected architectures.

illwieckz commented:
How do I build with MinGW for Windows on Linux? For Linux Arm?

How do I make a static `nacl_loader`? How do I rebuild with a 16k PageSize for Arm?

This scons stuff is over-convoluted…

illwieckz commented:
> Wrong. With Chromium `native_client` I can do `./scons --mode=nacl,opt-linux platform=arm sel_ldr`

I get that:

```
Exception: Cannot find a toolchain for arm in toolchain/linux_x86/pnacl_newlib_raw:
  File "SConstruct", line 2799:
    if UsingNaclMode(): nacl_env = nacl_env.Clone(
  File "/usr/lib/python3/dist-packages/SCons/Environment.py", line 1610:
    apply_tools(clone, tools, toolpath)
  File "/usr/lib/python3/dist-packages/SCons/Environment.py", line 117:
    _ = env.Tool(tool)
  File "/usr/lib/python3/dist-packages/SCons/Environment.py", line 2033:
    tool(self)
  File "/usr/lib/python3/dist-packages/SCons/Tool/__init__.py", line 265:
    self.generate(env, *args, **kw)
  File "site_scons/site_tools/naclsdk.py", line 756:
    _SetEnvForNativeSdk(env, root)
  File "site_scons/site_tools/naclsdk.py", line 109:
    raise Exception("Cannot find a toolchain for %s in %s" %
```

I have `arm-linux-gnueabihf-gcc` (and also `clang`).

slipher commented:
> How do I build with MinGW for Windows on Linux?

Found some documentation: https://github.com/DaemonEngine/native_client/blob/master/docs/build_systems.md. Although cross-architecture builds are supported, cross-OS builds are not. And you can't build with MinGW at all as it is designed for the MSVC toolchain. It seems building with MinGW would imply a porting effort beyond just the build system, as there is, e.g., a Microsoft assembler file.

The lack of cross-OS support would seem to be a limitation of Native Client though, not a limitation of Scons.

> For Linux Arm?

I posted that in the previous message.


> > Wrong. With Chromium `native_client` I can do `./scons --mode=nacl,opt-linux platform=arm sel_ldr`
> 
> With this exact command I get that:

You're getting an error finding a PNaCl toolchain, which should have been downloaded by `gclient` or whatever. Make sure you are in a fully equipped Chromium environment.

illwieckz commented:
> > > Wrong. With Chromium `native_client` I can do `./scons --mode=nacl,opt-linux platform=arm sel_ldr`
> > 
> > 
> > With this exact command I get that:
> 
> You're getting an error finding a PNaCl toolchain, which should have been downloaded by `gclient` or whatever. Make sure you are in a fully equipped Chromium environment.

Why do I need a PNaCl toolchain to build an Arm `sel_ldr`?

> The lack of cross-OS support would seem to be a limitation of Native Client though, not a limitation of Scons.

Yes, all that scons code in the repository is not meant for cross-compilation, that's what I meant.

> And you can't build with MinGW at all as it is designed for the MSVC toolchain. It seems building with MinGW would imply a porting effort beyond just the build system, as there is, e.g., a Microsoft assembler file.

Very annoying… I have seen they also have some Cygwin code, so I wonder if that can be used on MinGW, I haven't looked at this deeply though.

illwieckz commented:
Sorry, the output of `./scons --mode=opt-linux platform=arm sel_ldr` is (I forgot to remove the `nacl` mode):

```
AttributeError: 'SConsEnvironment' object has no attribute 'Program':
  File "SConstruct", line 3889:
    BuildEnvironments(selected_envs)
  File "site_init", line 198:
    
  File "/usr/lib/python3/dist-packages/SCons/Util/envs.py", line 242:
    return self.method(*nargs, **kwargs)
  File "site_scons/site_tools/defer.py", line 148:
    func(env)
  File "site_init", line 125:
    
  File "/usr/lib/python3/dist-packages/SCons/Script/SConscript.py", line 598:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/usr/lib/python3/dist-packages/SCons/Script/SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "src/trusted/validator_arm/build.scons", line 271:
    nexe = untrusted_env.ComponentProgram(test, 'testdata/' + test + '.S',
  File "/usr/lib/python3/dist-packages/SCons/Util/envs.py", line 242:
    return self.method(*nargs, **kwargs)
  File "site_scons/site_tools/component_builders.py", line 485:
    out_nodes = env.Program(prog_name, *args, **kwargs)
```

illwieckz commented:
This can already build `sel_ldr` on `linux-amd64`.

illwieckz commented:
This can now rebuild `sel_ldr` on `linux-armhf`.

illwieckz commented:
This can now rebuild `sel_ldr` on `linux-i686`.

illwieckz commented:
This can now rebuild `nacl_helper_bootstrap` on both `linux-amd64`, `linux-i686` and `linux-armhf`.

illwieckz commented:
This can now rebuild `sel_ldr` on `macos-amd64`.

illwieckz commented:
Using the same tricks, it is now possible to rebuild `sel_ldr` on for `windows-i686` with MinGW.

It means that it is now possible to rebuild the loader for all the DæmonEngine platforms using CMake, with macOS for the macOS loader, with Linux for all other platforms.

illwieckz commented:
In fact, only the single file `macamd64.inc` is enough. One can put it in `include-hax/winsdk/` and it works.

illwieckz commented:
Even better, those two obvious macros are enough:

```asm
push_reg macro Reg
	push Reg
	.pushreg Reg
	endm

alloc_stack macro Size
	sub rsp, Size
	.allocstack Size
	endm
```

So I provided a file named `include-hax/fake_masm/macamd64.inc` that just contains that, and when using MinGW, JWasm is used with `include-hax/fake_masm` as include directory.

illwieckz commented:
So this repository can now cross-compile `sel_ldr` for Windows using MinGW on Linux out of the box.

illwieckz commented:
I added a commit to statically link the `sel_ldr` binary, unfortunately it prints this warning:

```
transport_common.cc:(.text+0x78c): avertissement : Using 'gethostbyname' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

I don't know if that's a problem for us.

illwieckz commented:
Something I don't know how to do it, is to rebuild the armhf `sel_ldr` to use a 16K page size…

It still produces a 4K page size binary:

```
readelf -l src/trusted/service_runtime/sel_ldr | rg LOAD
  LOAD           0x000000 0x00010000 0x00010000 0x16c8e8 0x16c8e8 R E 0x1000
  LOAD           0x16d65c 0x0017d65c 0x0017d65c 0x0c054 0x21624 RW  0x1000
```

If at least one value on the right is `0X1000`, it is a 4K page size binary, they should all be `0x4000` or higher instead.

illwieckz commented:
> And you can't build with MinGW at all as it is designed for the MSVC toolchain.

@slipher In fact you cannot build with MSVC alone as it requires MinGW for compiling the `.S` files.
It requires both MSVC to assemble the `.asm` file and MinGW for assembing the `.S` files.

Proofs:

https://github.com/DaemonEngine/native_client/blob/a8ce8f617ede17a0771138d6deca8709778f8b73/SConstruct#L2345-L2354

With this CMake buildchain the Windows binary is also buildable on Linux using the system or use-provided MinGW and JWasm , while the SCons script relies on some Google-provided MinGW.

illwieckz commented:
So, I tested the rebuilt loader for `linux-amd64`.

Without any `nacl_helper_bootstrap`, it works either being shared or static (my system could already run the old loader without the boostrap):

```
./sel_ldr-shared -a -S -B irt_core.nexe -- helloworld-amd64.nexe 
DEBUG MODE ENABLED (bypass acl)
[1521736,2194020160:00:16:45.820317] BYPASSING ALL ACL CHECKS
[1521736,2194020160:00:16:45.820471] Native Client module will be loaded at base address 0x0000372b00000000
hello world

./sel_ldr-static -a -S -B irt_core.nexe -- helloworld-amd64.nexe 
DEBUG MODE ENABLED (bypass acl)
[1572450,529835072:00:20:05.601607] BYPASSING ALL ACL CHECKS
[1572450,529835072:00:20:05.601763] Native Client module will be loaded at base address 0x0000049600000000
hello world
```

But then, the rebuild bootstrap doesn't work and crashes:

```
./nacl_helper_bootstrap ./sel_ldr --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -Q -a -S -B irt_core.nexe -- helloworld-amd64.nexe 
Segmentation fault (core dumped)
```

If I use the old bootstrap with the new loader, it complains the dynamically-linked loader is not the way it expects it to be and aborts, and it complains the statically-linked loader is not dynamically-linked and aborts:

```
old/nacl_helper_bootstrap ./sel_ldr-shared --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -Q -a -S -B irt_core.nexe -- helloworld-amd64.nexe 
bootstrap_helper: ./sel_ldr: ELF file has unreasonable e_phnum=14

old/nacl_helper_bootstrap ./sel_ldr-static --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -Q -a -S -B irt_core.nexe -- helloworld-amd64.nexe 
bootstrap_helper: ./sel_ldr: ELF file not ET_DYN!  e_type=2
```

illwieckz commented:
`e_phnum` is likely the “_Number of program headers_”, I don't know what's that and why it would be a problem.

```
$ readelf -a sel_ldr-shared | grep 'Number of program headers'
  Number of program headers:         14
```

illwieckz commented:
The clang-built sel_ldr also has 15 program headers, and the source of `nacl_helper_bootstrap` has a max of 16…

```
$ readelf -a sel_ldr-clang | grep 'Number of program headers'
  Number of program headers:         14
```

The old booststrap prints the same error about `e_phnum` with the new loader (even if built with clang), so it looks like they bumped the max in source:

```
old/nacl_helper_bootstrap ./sel_ldr-clang --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -a -S -B irt_core.nexe -- helloworld-amd64.nexe 
bootstrap_helper: ./sel_ldr-clang: ELF file has unreasonable e_phnum=14
```

The new bootstrap built with clang can run the new loader built with GCC:

```
./nacl_helper_bootstrap-clang ./sel_ldr-gcc --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -a -S -B irt_core.nexe -- helloworld-amd64.nexe 
DEBUG MODE ENABLED (bypass acl)
[1818917,1943263040:00:46:45.463816] BYPASSING ALL ACL CHECKS
[1818917,1943263040:00:46:45.463958] Native Client module will be loaded at base address 0x00006f3800000000
hello world
```

illwieckz commented:
One good news is that the static `linux-amd64` loader runs the nexe on FreeBSD:

```
$ uname -s
FreeBSD
$ ./sel_ldr-static -Q -a -S -B irt_core.nexe -- helloworld-amd64.nexe
PLATFORM QUALIFICATION DISABLED BY -Q - Native Client's sandbox will be unreliable!
DEBUG MODE ENABLED (bypass acl)
[1385,7976000:22:54:24.309245] BYPASSING ALL ACL CHECKS
[1385,7976000:22:54:24.309867] Native Client module will be loaded at base address 0x0000617200000000
hello world
```

It means the requirement for running the game on FreeBSD would just be to start the linuxulator service, without requiring to install any system package to get a linux libc!

illwieckz commented:
So, now, when the compiler is not a Clang-derivative, it build the `nacl_helper_bootstrap` binary with Clang in a subproject (like we do with NaCl VMs), because `nacl_helper_bootstrap` only works when built with Clang.

illwieckz commented:
So, it looks like the `ET_DYN` check in `nacl_helper_bootstrap` was just some lazy check for the file being in executable format. Adding another check for `ET_EXEC` makes it possible to run statically build `sel_ldr`!

```
$ file sel_ldr
sel_ldr: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, BuildID[sha1]=cbf906bedf2088500b220f244900ad9c94399fbf, for GNU/Linux 3.2.0, not stripped

$ ./nacl_helper_bootstrap ./sel_ldr --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -a -S -B irt_core.nexe -- helloworld-amd64.nexe
DEBUG MODE ENABLED (bypass acl)
[306836,200389696:02:55:52.963842] BYPASSING ALL ACL CHECKS
[306836,200389696:02:55:52.964008] Native Client module will be loaded at base address 0x0000039700000000
hello world
```

illwieckz commented:
Now, what's remaining on the list of things I would like to see being implemented is to get a 16k PageSize `sel_ldr` binary for `linux-armhf`.

illwieckz commented:
So, I found a way to build a `sel_ldr` for `linux-armhf` that uses 16K PageSize:

```
$ readelf -l sel_ldr | grep LOAD
  LOAD           0x000000 0x00000000 0x00000000 0xf9704 0xf9704 R E 0x4000
  LOAD           0x0fb980 0x000ff980 0x000ff980 0x08f4c 0x1b6c8 RW  0x4000
```

But the `nacl_helper_bootstrap` is still using a 4K PageSize:

```
$ readelf -l nacl_helper_bootstrap | grep LOAD
  LOAD           0x000000 0x00010000 0x00010000 0x02666 0x02666 R E 0x1000
  LOAD           0x000000 0x00013000 0x00013000 0x00000 0x01005 RW  0x1000
  LOAD           0x000000 0x00015000 0x00015000 0x3ffed000 0x3ffed000     0x1000
  LOAD           0x003000 0x40002000 0x40002000 0x00014 0x00014 RW  0x1000
```

illwieckz commented:
So, I found a way to build `nacl_helper_bootstrap` for `linux-armhf` with 16K PageSize:

```
$ readelf -l nacl_helper_bootstrap | grep LOAD
  LOAD           0x000000 0x00010000 0x00010000 0x02666 0x02666 R E 0x4000
  LOAD           0x000000 0x00014000 0x00014000 0x00000 0x01005 RW  0x4000
  LOAD           0x002000 0x00016000 0x00016000 0x3ffec000 0x3ffec000     0x4000
  LOAD           0x006000 0x40002000 0x40002000 0x00014 0x00014 RW  0x4000
```

I have not tested it though, and I wonder if the armhf nexe also have to be built with a 16K PageSize too.

illwieckz commented:
To run on 16K PageSize kernel, it looks like we would also need to rebuild the nexe as well:

```
$ readelf -l irt_core-armhf.nexe | grep LOAD
  LOAD           0x010000 0x0ffc0000 0x0ffc0000 0x30000 0x30000 R E 0x10000
  LOAD           0x000000 0x3efe0000 0x3efe0000 0x041b8 0x041b8 R   0x10000
  LOAD           0x0041b8 0x3eff41b8 0x3eff41b8 0x00b90 0x011b0 RW  0x10000
```

slipher commented:
The nexe is loaded by sel_ldr, not the system executable loader right? So maybe its "page size" doesn't matter, or maybe the code of sel_ldr has to be changed. Hard to speculate since I don't understand the concept of page size of an executable.

slipher commented:
https://tristanxr.com/post/why-16k-page-size/ has more details on Asahi Linux's experiences with page size portability issues. It says that most Linux programs work without issues. The ones with problems are ones that manage their own memory mappings somehow, e.g. using a custom allocator instead of libc's. In the thing about Windows games you linked, the problem is probably with the Windows memory allocator or some other part of the Windows runtime. The problem is surely not with executable alignment: given that you are translating to a different ISA, you can lay out the translated instructions however you want. So I don't think the section alignments in the program header really matter for 16k page compatibility. As a last bit of evidence, I tried setting a binary's executable section's alignment to 1, and it still worked.

I do very much expect that sel_ldr is a program that manages its own memory mappings. Firstly, I believe that it maps the nexe into memory itself, rather than using the kernel's executable loading. Secondly, The NaCl sandbox model specifies a specific range of memory that the untrusted code is allowed to read/write access. All memory allocations by untrusted code must go in this range and all allocations by trusted code outside it. If there is an NaCl syscall for mapping more memory pages, we should check whether the page size is a hard-coded number, baked in at compile time from a system header, or queried at runtime.

Note that if you read the nexe program header dump closely, the alignment is not 4k, but *64k*. So no problem even if it somehow mattered.

illwieckz commented:
The `nacl_helper_bootstrap` for `linux-armhf` crashes when the `sel_ldr` is linked statically.

It works on `linux-amd64`, but not on `linux-armhf`… That's annoying because arm is the platform where it would be easier to have a static `sel_ldr` (less files to ship, and a simpler loading chain).

Even when building with debug symbols, I cannot debug the tool:

```
Reading symbols from ./nacl_helper_bootstrap...
(gdb) r
Starting program: ./nacl_helper_bootstrap ./sel_ldr --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -a -S -B irt_core.nexe -- helloworld-armhf.nexe

Program received signal SIGSEGV, Segmentation fault.
0xf7f34924 in ?? ()
(gdb) bt
#0  0xf7f34924 in ?? ()
#1  0xf7ea78c0 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

slipher committed:
Fix import error on Python 3.13+

'pipes' module was removed from the standard library.

slipher committed:
Fix obsolete Python unittest assert aliases

assertEquals and assertNotEquals were removed in Python 3.12.

slipher committed:
Mark run_sel_ldr_seccomp_test broken

Doesn't work on my Debian Trixie. Anyway I believe this seccomp BPF
thing is an experiment that was never really productionized.

After this everything passes for me with
./scons --mode=opt-host,nacl saigo=1 platform=x86-64 all_tests

slipher committed:
Fix more SyntaxWarning's in build system

slipher committed:
Fix compile error in nacl_signal.c

slipher committed:
run.py: don't assume directory name

This fixes some test (if the repository dir is not named exactly
native_client).

slipher committed:
Check in and use old Googletest

Check in a copy of the 2013-vintage Googletest which is specified in the
DEPS file. Trying to use an external one would not work because an old
version that supports old C++ standards is needed. Hook it up to the
build system so that most tests work now.

Imported from gtest revision 2d3543f81d6d4583332f8b60768ade18e0f96220:
- LICENSE
- src/
- *.h files within include/

slipher committed:
Fix run.py shebang

slipher committed:
Don't bogusly disable minidump tests

Clearly the test shouldn't be disabled with Saigo for that condition.
But I couldn't find any problem with running the test for any PNaCl
configuration either (it always had a build id).

slipher committed:
Make crash dump tests runnable

Like upstream, you must pass a breakpad_tools_dir argument for the tests
to run. Unlike upstream, the Breakpad #includes will be looked for also
in the tools dir, instead of the Chromium code depot.

slipher committed:
Disable a test that uses PNaCl even in Saigo mode

illwieckz committed:
cmake: fix building nacl_helper_bootstrap with zero-based 64-bit sandbox with Clang

illwieckz committed:
cmake: fix building nacl_helper_bootstrap with zero-based 64-bit sandbox with Clang

slipher committed:
Don't use --sysroot flag by default

For Linux toolchains (except x86-64 where I previously commented it
out), there were hard-coded sysroot flags referring to the paths of
Chromium's bundled dependencies. Get rid of this. On my machine the
cross toolchains work fine without it. But I added a
a 'sysroot' command line argument, so if anyone needs it they can
pass sysroot=... on the scons command line to configure a --sysroot flag.

illwieckz committed:
cmake: use size optimization for nacl_helper_bootstrap, with toplevel-reorder disabled on armhf

The -ftoplevel-reorder option breaks the build for armhf.

illwieckz committed:
cmake: use size optimization for nacl_helper_bootstrap, with toplevel-reorder disabled on armhf

The -ftoplevel-reorder option breaks the build for armhf.

slipher committed:
test_lib Python 3 str/bytes workaround

This one only seems to be triggered when a test fails.

slipher committed:
Make ARM tests runnable with QEMU

Use system qemu-armhf for the emulator for tests. The command here
basically matches what the run_under_qemu_arm script (which can be seen
at tools/trusted_cross_toolchains/qemu_tool_arm.sh) was doing.

Also change the emulator in run.py.

slipher committed:
ARM: fix NaClSysTestInfoLeak misaligned load

This fixes a NaCl syscall that only exists for a test. Fixes the
run_infoleak_test test target on ARM.

slipher committed:
Fix REGS_SAVER_FUNC test util on ARM

REGS_SAVER_FUNC is meant to work even if the stack pointer is
misaligned, but the multi-register PUSH instructions crash if the stack
is misaligned, as per the ARM documentation. I guess the old(?) bundled
QEMU used by upstream doesn't implement this faithfully. Fix it by using
a series of single-register pushes instead.

Fixes the run_register_set_test test target on ARM.

illwieckz commented:
Now fixed.

That PR looks ready now.

necessarily-equal commented on src/trusted/platform_qualify/arch/x86/vcpuid.c:
```diff
@@ -386,13 +386,13 @@ static int asm_HasCX8(void) {
 #endif  /* 64-bit Windows */
 
-#if NACL_WINDOWS && (NACL_BUILD_SUBARCH == 64)
+#if !(NACL_WINDOWS && (NACL_BUILD_SUBARCH == 64)) || !defined(MSC_VER)
```
This here inverts the condition in a way that doesn't look intentional and doesn't match the comment below.

illwieckz commented:
I removed everything that wasn't CMake, other things like the MinGW compatibility code will be submitted in a separate PR.

illwieckz commented:
The CMake code already includes stuff for MinGW:

- https://github.com/DaemonEngine/native_client/pull/7

So this part references files that aren't there yet.

slipher commented:
> I removed everything that wasn't CMake, other things like the MinGW compatibility code will be submitted in a separate PR.

Thanks. I will continue to use the Scons build since I am interested in running the tests, among other things. So I won't have to review this super closely since it is only CMake files which can't break the other build.

Are you really using all those Daemon `*.cmake` files, like `Compiler.cmake` and `SourceGenerator.cmake`? I don't see any use of `daemon_embed_files` for example.

slipher committed:
Fix GCC warnings for code that intentionally segfaults

Use `volatile` as optimization barrier for the value of the pointer (NOT
for what's pointed at) like in Daemon's /injectFault segfault.

fixup: crashing

slipher committed:
Fix GCC warning by checking snprintf result

slipher committed:
Work around GCC function pointer cast warnings

slipher committed:
Fix GCC warning about macro string literal concat

slipher committed:
Fix GCC fclose warning in a test

This test is using some custom macros instead of googletest and
ASSERT_EQ apparently double-evaluates the arguments if the assertion
fails.

slipher committed:
Discard extra sections linking nacl_helper_bootstrap

Use --orphan-handling=unwanted when linking nacl_helper_bootstrap which
makes the linker discard any sections not explicitly mentioned in the
linker script. This prevents the linker from producing a defective
binary with overlapping PT_LOAD headers when using GCC on certain
distros such as Ubuntu and Arch that have it configured to produce a
.note.gnu.property section

illwieckz committed:
cmake: discard extra sections linking nacl_helper_bootstrap

See d281e6a3f08a13a7c7ba6ffd7160b2267a3c8765

illwieckz committed:
cmake: discard extra sections linking nacl_helper_bootstrap

See d281e6a3f08a13a7c7ba6ffd7160b2267a3c8765
